functions/slack_slash_command/index.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_slack_setup]
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use GuzzleHttp\Psr7\Response;
+
+require_once './vendor/autoload.php';
+// [END functions_slack_setup]
+
+// [START functions_verify_webhook]
+/**
+* Verify that the webhook request came from Slack.
+*/
+function isValidSlackWebhook(ServerRequestInterface $request): bool
+{
+    /**
+    * PHP Functions Framework does not support "global"/instance-scoped
+    * constants, so we fetch these values within PHP functtions themselves.
+    */
+    $SLACK_SECRET = getenv("SLACK_SECRET");
+
+    // Check for headers
+    $timestamp = $request->getHeader("X-Slack-Request-Timestamp");
+    $signature = $request->getHeader("X-Slack-Signature");
+    if (!$timestamp || !$signature) {
+        return false;
+    } else {
+        $timestamp = $timestamp[0];
+        $signature = $signature[0];
+    }
+
+    // Compute signature
+    $plaintext = 'v0:' . $timestamp . ':' . (string) $request->getBody();
+    $hash = 'v0=' . hash_hmac('sha256', $plaintext, $SLACK_SECRET);
+
+    return $hash === $signature;
+}
+// [END functions_verify_webhook]
+
+// [START functions_slack_format]
+/**
+* Format the Knowledge Graph API response into a richly formatted Slack message.
+*/
+function formatSlackMessage($kgResponse, $query): string
+{
+    $responseJson = [
+        "response_type" => "in_channel",
+        "text" => "Query: " . $query
+    ];
+
+    $entityList = $kgResponse["itemListElement"];
```

Comment:
Does $kgResponse have a failure mode which is not a thrown exception but for which itemListElement is not populated?

```
PHP Notice:  Undefined index: itemListElement in php shell code on line 1
```

This warning can be suppressed with a check or a default value, such as a null coalescing operator.

```suggestion
    $entityList = $kgResponse["itemListElement"] ?? null;
```

Reply:
I believe `itemListElement` is supposed to be defined (but empty!) _even if_ there are no matching results. (There _is_ a test case for this behavior.)

Happy to add a null-coalescing operator for defensive programming though (e.g. to mitigate errors and/or unexpected changes at the API level), if you think that's worth doing.